### DISCUSSION TITLE
@Transactional 전파 속성에 대해 설명해 주세요

### DISCUSSION THREAD

**Discussion by 3Juhwan** [2024-woowacourse-study/level-interview#228] [Spring Data]
.

**Comment by nayonsoso** [comment]
### 물리 트랜잭션과 논리 트랜잭션

#### 물리 트랜잭션
- 실제 데이터베이스에 적용되는 트랜잭션
- Connection 을 통해 커밋/롤백하는 단위

#### 논리 트랜잭션
- 스프링이 트랜잭션 매니저를 통해 트랜잭션을 처리하는 단위

스프링은 다음과 같은 규칙으로 트랜잭션을 처리한다.
- 모든 논리 트랜잭션이 커밋되어야 물리 트랜잭션이 커밋됨
- 하나의 논리 트랜잭션이라도 롤백되면 물리 트랜잭션은 롤백됨

### 트랜잭션 전파 속성이란
이미 트랜잭션이 진행중일 때, 추가 트랜잭션 진행을 어떻게 할지 결정하는 것이다. 
기존 트랜잭션에 참여할 수도 있고, 별도의 트랜잭션으로 진행할 수도 있고, 에러를 발생시키는 등의 선택을 할 수 있다. 
이때 ‘참여한다’는 것은 둘 이상의 논리 트랜잭션이 하나의 물리 트랜잭션으로 처리되는 것을 의미한다.
@Transactional 어노테이션에서 propagation 옵션으로 설정할 수 있다.
cf. 새로운 논리 트랜잭션을 ‘외부 트랜잭션’, 기존 논리 트랜잭션을 ‘내부 트랜잭션’ 이라고도 한다.

#### REQUIRED
- 기본 전파 속성
- 트랜잭션이 필요하다는 뜻이다(이래서 ‘required 인듯).
- 기존 트랜잭션이 있으면 거기에 참여하고, 없으면 새로 만든다.

#### REQUIRES_NEW
- 항상 새로운 트랜잭션이 필요하다는 뜻이다.
- 기존 트랜잭션이 있든 없든 새로운 물리 트랜잭션을 만든다.

#### SUPPORTS
- 트랜잭션이 있으면 지원하겠다는 뜻이다.
- 기존 트랜잭션이 없다면 없는대로 진행하고, 있다면 참여한다.

#### MANDATORY
- 트랜잭션이 반드시 필요하다는 뜻이다.
- 기존 트랜잭션이 없다면 예외를 발생시키고, 있다면 참여한다.

#### NOT_SUPPORTED
- 트랜잭션을 지원하지 않는다는 뜻이다.
- 기존 트랜잭션이 있어도 없어도 트랜잭션 없이 진행한다.

#### NEVER
- 트랜잭션을 허용하지 않겠다는 뜻이다.
- 기존 트랜잭션이 있으면 예외를 발생시킨다.

#### NESTED
- 중첩 트랜잭션을 만들겠다는 뜻이다.
- 기존 트랜잭션이 없으면 새로운 트랜잭션을 만들고, 있으면 그 트랜잭션의 자식 트랜잭션을 만든다.

<img src= "https://img1.daumcdn.net/thumb/R1280x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2FbHIrEE%2FbtrILdW6dPy%2FxCwRpCG9RzXDlhfdJmjOY0%2Fimg.png"  /> 
출처: https://mangkyu.tistory.com/269 [MangKyu's Diary:티스토리]

**Comment by gyungchan-jo** [comment]
`@Transactional` 어노테이션은 트랜잭션 관리에 사용되며, 트랜잭션의 전파(Propagation) 속성은 트랜잭션이 어떻게 전파될지를 결정합니다. Spring에서는 여러 가지 전파 속성을 제공하며, 각각의 속성은 트랜잭션 경계 내에서의 동작 방식을 정의합니다. 주요 전파 속성들은 다음과 같습니다:

1. REQUIRED (기본값)
: 현재 트랜잭션이 존재하면 이를 사용하고, 존재하지 않으면 새로운 트랜잭션을 생성합니다.

2. REQUIRES_NEW
: 항상 새로운 트랜잭션을 생성하고, 기존 트랜잭션이 존재하면 일시 중단합니다.

3. SUPPORTS
: 현재 트랜잭션이 존재하면 이를 사용하고, 존재하지 않으면 트랜잭션 없이 실행합니다.

4. NOT_SUPPORTED
: 현재 트랜잭션이 존재하면 이를 일시 중단하고, 트랜잭션 없이 실행합니다.

5. MANDATORY
: 현재 트랜잭션이 존재해야 하며, 존재하지 않으면 예외를 발생시킵니다.

6. NEVER
: 현재 트랜잭션이 존재하면 예외를 발생시키고, 존재하지 않으면 트랜잭션 없이 실행합니다.

7. NESTED
: 현재 트랜잭션이 존재하면 중첩된 트랜잭션을 생성하고, 존재하지 않으면 REQUIRED와 동일하게 동작합니다.
